### PR TITLE
Generalizing interface for PlaneSmoother

### PR DESCRIPTION
**Pull request by ScottMacLachlan**
This tries to address #1973, allowing a more general interface for PlaneSmoother to specify which set of points to use to divide an axis to define the PlaneSmoother patch.  Now, if the user gives a number in the usual `sweep` format (e.g., `0+50`), then the old behaviour is preserved, with evenly spaced bins.  If, instead, the user gives a key (e.g., `0+key`), then that key is looked for in the `appctx` and used as `divisions` in place of the evenly spaced bins.

### CODE REVIEW

File: firedrake/preconditioners/patch.py [hunk 1]
```diff
@@ -585,10 +585,13 @@ def coords(dm, p, coordinates):
         bary /= len(closure_of_p)
         return bary
 
-    def sort_entities(self, dm, axis, dir, ndiv):
+    def sort_entities(self, dm, axis, dir, ndiv=-1, divisions=None):
```
Comment (dham) [inline firedrake/preconditioners/patch.py:588]: Use `None` as the out of band value, not `-1`

**Review by dham — CHANGES_REQUESTED** [review]

File: firedrake/preconditioners/patch.py (hunk 1 above)
Comment (ScottMacLachlan) [reply firedrake/preconditioners/patch.py]: Done.

**Comment by ScottMacLachlan** [comment]
Alright - my apologies for leaving this sitting for so long, but ...  I think it's ready for a proper review now.

This now includes two enhancements to PlaneSmoother, the original one (allowing more general specification of the spacing of bins into which to divide points) and a second one allowing more general selection of how points are mapped into those bins.  

Two of the tests check correctness of these against the original syntax for PlaneSmoother, while the other two check reference performance for their intended use cases, on non-uniform tensor-product meshes, or for line/plane relaxation that doesn't align with the coordinate axes...

**Review by dham — CHANGES_REQUESTED** [review]
Update to current master and check that this passes tests, and we're good.

**Comment by ScottMacLachlan** [comment]
Update to master seems to have gone smoothly...

**Review by dham — APPROVED** [review]

**Pull request merged**